docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,663 @@
+{
+    int getValue();
+}
+
+void print(each T)(expand each T) {...}
+
+void f<each T : IFloat>(expand each T v)
+{
+    print(expand (each v).getValue());
+}
+```
+
+Here, `expand (each v).getValue()` will expand the pattern expression `(each v).getValue()` into a pack of values. The result of this `expand` expression
+is a pack of values where each element of the pack is computed by substituting `each v` in the pattern expression with each element in `v`. The resulting
+pack of `int` values is then passed to `print` function that also takes a pack of values.
+
+For now, we require that all variadic generic type packs to appear in the end of a parameter list, after any ordinary parameters. This means that the following
+definitions are invalid:
+
+```
+void f<each T, U>() {} // Error, ordinary parameter `U` after type pack.
+void g<each T, U = int>() {} // Error, ordinary parameter after type pack.
+void k<each T, let i : int>() {} // Error.
+void h<U = int, each T>() {} // OK.
+```
+
+Addtionally, we establish these restrictions on how `expand` and `each` maybe used:
+- The pattern type of an `expand` type expression must capture at least one generic type pack parameter in an `each` expression.
+- The type expression after `each` must refer to a generic type pack parameter, and the `each` expression can only appear inside an `expand` expression.
+
+These rules means that expressions like `expand int`, or `each T` on its own are invalid expressions.
+
+Similarly, when using `expand` and `each` on values, we require that:
+- The pattern expression of an `expand` expression must capture at least one value whose type is a generic type pack parameter.
+- The expression after `each` must refer to an value whose type is a generic type pack parameter, and the `each` expression can only appear inside an `expand` expression.
+
+Detailed Explanation
+--------------------
+
+To implement variadic generics, we need to introduce several semantic constructs in our type system.
+### `GenericTypePackParameterDecl`
+When a generic parameter is defined with the `each` keyword, such as in `void f<each T>`, the parser should create a new type of AST node inside the generic, and we name this
+AST node a `GenericTypePackParameterDecl`. With this additional, a generic parameter can be `GenericTypeParameterDecl`, `GenericValueParameterDecl`, `GenericTypeConstraintDecl`
+and `GenericTypePackParameterDecl`. When the user defines type constraints on a generic type pack parameter, we will form a `GenericTypeConstraintDecl` whose `subType` is a
+`DeclRefType` referencing the `GenericTypePackParameterDecl`.
+
+### Type Pack
+A type pack represents a pack of types. The simplest form of a type pack is a `ConcreteTypePack` that is a list of concerete type packs, such as `(int, float, float3)`.
+In a generic decl such as `void f<each T>(T v)`, `T` refers to an abstract type pack represented by the generic type pack parameter `T`. The type of parameter `v` in this case
```

Comment:
`void f<each T>(T v)`

I get this is just an example, but is it only valid when where is `expand each` keyword in the parameter list?
e.g. `void f<each T>(expand each T v)`

Reply:
using `T` directly here is fine, it also means a pack. `expand each T` is strictly equivalent to `T`.

Reply:
I see the simplification rule below.